libs/api/database/src/lib/content/stores/tags.store.ts

```diff
@@ -0,0 +1,167 @@
+import { TagsTree } from '@dragonfish/shared/models/content/tags.model';
+
+@Injectable()
+export class TagsStore {    
+    private readonly logger: Logger = new Logger(TagsStore.name);
+
+    constructor(
+        @InjectModel('Tags') private readonly tags: Model<TagsDocument>,
+    ) {}
+
+    /**
+     * Returns all tags of a specified kind, sorted alphabetically.
+     * @param kind
+     */
+    async fetchTags(kind: TagKind): Promise<TagsDocument[]> {
+        return this.tags.find({ kind: kind }).sort({ name: 1 });
+    }
+
+    /**
+     * Returns a single tag with the given ID, or null if not found.
+     * @param id The ID of the tag to find.
+     * @returns The tag with the ID.
+     */
+    async findOne(id: string): Promise<TagsDocument> {
+        return this.tags.findById(id);
+    }
+
+    /**
+     * Returns the given tag and all its descendants.
+     * @param king
+     * @param tagId
+     */
+    async fetchDescendants(tagId: string): Promise<TagsTree> {
+        const parentTag = await this.tags.findById(tagId);
+        if (!parentTag) {
+            throw new NotFoundException("No tag with that ID found.");
+        }
+
+        const root: TagsTree = {
+            ...parentTag.toObject(),
+            children: []
+        };
+        const tree = await this.buildDescendantTree(root);
+
+        return tree;
+    }
+
+    private async buildDescendantTree(tagNode: TagsTree): Promise<TagsTree> {
+        const currentRoot = await this.populateImmediateChildren(tagNode._id);
+        if (!currentRoot.children || currentRoot.children.length === 0) {
+            // If we can't find any children for this node, it is a leaf. Return it.
+            return tagNode;
+        }
+
+        // Otherwise, iterate through each child node, find _its_ children,
+        // and add them to the node's `children` property.
+        for (const child of currentRoot.children) {
+            child.children = (await this.buildDescendantTree(child))?.children;
+        }
+
+        return currentRoot;
+    }
+
+    /**
+     * Returns the Tag, with an extra `children` property array that contains all of this
+     * tag's immediate children.
+     * @param id The ID of the tag to populate.
+     */
+    private async populateImmediateChildren(id: string): Promise<TagsTree> {
+        // We force maxDepth to be 0 so that we only get the node's immediate children.
+        // If we went further, Mongo would just give us a flat array, and we'd have to
+        // put it together ourselves anyway.
+        const results: TagsTree[] = await this.tags.aggregate<TagsTree>([
+            { $match: { _id: id } }
+        ])
+        .graphLookup({
+            from: 'tags',
+            startWith: '$_id',
+            connectFromField: '_id',
+            connectToField: 'parent',
+            as: 'children', // Careful: this string should match the name of the property in `TagsTree`.
+            maxDepth: 0,
```

Comment:
`graphLookup` is some fuckin' voodoo magic. seriously, check out how it works. I'm force-limiting it to a `maxDepth` of 0 and building the child tree by hand, because it gives us back a flat list by default, which ain't all that useful.

But if all we needed was a flat list, it's the tool for the job

Reply:
interesting, that'll be useful in the future